source/transactions/tests/legacy/mongos-pin-auto-tests.py

```diff
@@ -16,7 +16,7 @@
 runOn:
     -
         minServerVersion: "4.1.8"
-        topology: ["sharded"]
+        topology: ["sharded", "load-balanced"]
```

Comment:
Adding "load-balanced" here and in `pin-mongos.yml` may be problematic, as I observed some test failures in PHPLIB (see: [task failure](https://parsley.mongodb.com/test/mongo_php_library_test_loadBalanced__php_edge_versions~latest_stable_mongodb_versions~6.0_driver_versions~latest_stable_os~debian11_test_loadBalanced_patch_f6ce2119fad23641df100248c11f081f15ad6cb5_64f22ec72fbabe969b92bee1_23_09_01_18_35_31/0/4b0564721909a64a107218d00faa4e06?bookmarks=0,8&shareLine=1)).

This could be related to the LB spec defining separate rules for [connection pinning](https://github.com/mongodb/specifications/blob/master/source/load-balancers/load-balancers.rst#connection-pinning), which might not be compatible with the assumptions made in these tests for normal [mongos pinning](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#mongos-pinning).

@jyemin: Any insights here? Does Java have any problem passing these pinning tests on LB and/or serverless?

Reply:
I'm pretty sure the assertion failure on PHP here is by design, as we use [`mongoc_client_session_get_server_id()`](https://mongoc.org/libmongoc/current/mongoc_client_session_get_server_id.html) to check for a pinned session, and a LB topology only has a single server so it's unlikely that method would return anything. Furthermore, PHP uses libmongoc's single-threaded mode, so we are implicitly pinned to a single connection at all times (LB spec's rules for [connection pinning](https://github.com/mongodb/specifications/blob/master/source/load-balancers/load-balancers.rst#connection-pinning) wouldn't even come into play).

That said, I'm curious if these tests might still apply to other languages, in which case I should only skip them in PHP's test suite (and libmongoc can do the same if running in single-threaded mode).

Reply:
Additionally, several other pinning tests fail due to [`recoveryToken` fields](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#recoverytoken-field) not being found in outgoing `commitTransaction` and `abortTransaction` commands (see: [task failure](https://parsley.mongodb.com/test/mongo_php_library_test_loadBalanced__php_edge_versions~latest_stable_mongodb_versions~6.0_driver_versions~latest_stable_os~debian11_test_loadBalanced_patch_f6ce2119fad23641df100248c11f081f15ad6cb5_64f22ec72fbabe969b92bee1_23_09_01_18_35_31/0/7589ac969195a0bb11acdccb96538e4a?bookmarks=0,7&shareLine=1)).

Reply:
@jyemin: I confirmed with @kevinAlbs that libmongoc intentionally does not pin LB connections, since a client would only have one connection to each server (libmongoc implements client pooling instead of CMAP-compliant connection pooling). Given that, I think I'll just need to skip these tests for PHP.

I'll defer to you to confirm that all of these pass for Java, though.

Reply:
I think I ran all these tests when I first reviewed the PR and the only failures were the five that failed on Serverless.